File: src/csharp/api/Models/BaseEntity.cs
namespace Example.Api.Models;

/// <summary>
/// Base entity with common properties.
/// </summary>
/// <value></value>
public abstract record BaseEntity
{
    /// <summary>
    /// The date and time when the order was created.
    /// </summary>
    public DateTimeOffset CreatedAt { get; set; }

    /// <summary>
    /// The date and time when the order was last updated.
    /// </summary>
    /// <value></value>
    public DateTimeOffset UpdatedAt { get; set; }
}

File: src/nodejs/src/entities/BaseEntity.ts
import { CreateDateColumn, UpdateDateColumn, ValueTransformer } from "typeorm";

/**
 * Transformer to ensure dates are always stored and retrieved as UTC.
 */
const utcDateTransformer: ValueTransformer = {
  to: (value: Date | undefined): Date | undefined => {
    return value;
  },
  from: (value: string | Date | undefined): Date | undefined => {
    return value ? new Date(value) : undefined;
  },
};

/**
 * Base entity class providing common timestamp fields for all entities.
 */
export abstract class BaseEntity {
  /**
   * The timestamp when the entity was created (stored as UTC in database).
   */
  @CreateDateColumn({
    type: "timestamptz",
    name: "created_at",
    transformer: utcDateTransformer,
  })
  createdAt!: Date;

  /**
   * The timestamp when the entity was last updated (stored as UTC in database).
   */
  @UpdateDateColumn({
    type: "timestamptz",
    name: "updated_at",
    transformer: utcDateTransformer,
  })
  updatedAt!: Date;
}
